[mono][wasm] Add a minimal version of gsharedvt for llvmonly, enabled…

vargaz opened the pull request:
… by default.

In this version, variable length arguments/locals are not supported. This means
that these compiled methods don't have the overhead of gsharedvt, i.e. wrappers,
dynamically allocated/copied locals, etc. They have similar performance to normal
generic sharing.

This can handle methods like List<T>.get_Item (), System.GC:AllocateArray<T>, etc.

vargaz committed:
[mono][wasm] Add a minimal version of gsharedvt for llvmonly, enabled by default.

In this version, variable length arguments/locals are not supported. This means
that these compiled methods don't have the overhead of gsharedvt, i.e. wrappers,
dynamically allocated/copied locals, etc. They have similar performance to normal
generic sharing.

This can handle methods like List<T>.get_Item (), System.GC:AllocateArray<T>, etc.

vargaz commented:
/azp run runtime-wasm

vargaz commented:
/azp run runtime-wasm

lambdageek commented on src/mono/mono/mini/mini.c:
```diff
@@ -3261,6 +3261,18 @@ mini_method_compile (MonoMethod *method, guint32 opts, JitFlags flags, int parts
+			 * Enable a minimal version of gsharedvt where only
+			 * methods which don't really depend on their generic
+			 * arguments are supported, like List<T>.get_Count ().
+			 * - no signatures or locals with variable length types
```
Would it be possible to extend this so that if the function has a local or argument of generic instance of a reference type e.g. `List<T>` - those aren't really variable sized.

You're already sort of supporting this because `this` "uses" the gsharedvt generic param.


Actually, do we need to check the type of `this` in `mini_is_gsharedvt_signature`?  it's always boxed if we had a generic struct, right?

vargaz replied:
Fixed.

vargaz commented:
The test failures are relevant.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Fix gsharedvt returns on wasm.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Allow byref gsharedvt types in gsharedvt in/out wrappers.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Avoid setting both ctx->values and ctx->addresses for arguments of type LLVMArgGsharedvtFixedVtype.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Fix support for gsharedvt+deopt.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Disable UNBOX_ANY in gsharedvt_min mode.

vargaz commented:
/azp run runtime-wasm

vargaz committed:
Fix llvmonly support for static methods with bound NULL first argument.

Some code paths can end up calling the general delegate invoke wrapper, which used
target!=NULL to decide whenever the call is bound, but the bound argument can be NULL.
So add a 'bound' field to Delegate and check it in the wrapper.

Some code paths might not set 'bound' so retain the target!=NULL check as well.

vargaz commented:
/azp run runtime-wasm
